Array Scopes, Error Event Definitions & Scoped Error Bindings

pinussilvestrus opened the pull request:
Closes #20

* handle array scopes
* allow error event definitions
  * validate for missing errorRef
  * validate for property type ("String", "Hidden", "Dropdown")
* allow error scopes
  * validate for missing id

MaxTru commented on resources/schema.json:
```diff
@@ -467,6 +467,52 @@
+    "scopes": {
+      "$id": "#/definitions/scopes",
+      "type": "array",
+      "description": "Special scoped bindings that allow you to configure nested elements",
```
Any reasons why we dont have a `title` property for `scopes` (we have it for the other cases)

MaxTru commented on resources/schema.json:
```diff
@@ -467,6 +467,52 @@
+      "$id": "#/definitions/scopes",
+      "type": "array",
+      "description": "Special scoped bindings that allow you to configure nested elements",
+      "items": {
```
Also for `items` => why no `title` + why no `description`?

MaxTru commented on resources/schema.json:
```diff
@@ -467,6 +467,52 @@
+          "type": {
+            "$id": "#scopes/item/type",
+            "type": "string",
+            "pattern": "^(camunda|bpmn):"
```
We only allow scopes of type `camunda:Connector` or `bpmn:Error` right? Why not check explicitly for these two?

MaxTru commented on test/spec/validationSpec.js:
```diff
@@ -23,6 +23,10 @@ describe('validation', function() {
 
   function testTemplate(name, file, only = false) {
 
+    if (!file) {
```
Nice!

MaxTru reviewed (commented):
Really nice how well we can extend the schema - JSONSchema is powerful :-)

Only minor comments, apart these.

:rocket:

pinussilvestrus commented:
Thanks for the review, very good points 👍  I added a comment on top (https://github.com/camunda/element-templates-json-schema/pull/21/commits/c5d310fa75e998980ce71e23fb5c216b4a6d21de) which changes the pattern with a more restrictive rule. Also added the missing `title` & `description` fields.

MaxTru reviewed (approved):
:clap:
